ENH: Streaming-equivalence GTest for GaussianInterpolateImageFunction

romangrothausmann opened the pull request:
Adds a GTest regression guard for [issue #1011](https://github.com/InsightSoftwareConsortium/ITK/issues/1011), which was fixed by #1202: `ResampleImageFilter` + `GaussianInterpolateImageFunction` must produce identical output whether or not its downstream is streamed.

Force-pushed to revive this dormant draft from 2019. Original three commits superseded by a single modern GTest TEST block; authorship preserved via `Co-Authored-By:` trailer.

<details>
<summary>Why the original draft no longer reproduced</summary>

The pipeline in the original draft fed the resampler an in-memory `Image` object directly. An in-memory image's `BufferedRegion` always equals its `LargestPossibleRegion` regardless of downstream `RequestedRegion` propagation, so the bug surface in `GaussianInterpolateImageFunction::ComputeBoundingBox` (which read from `BufferedRegion` pre-#1202) never triggered.

The new TEST inserts a `CastImageFilter` upstream of the resampler — a streaming-aware no-op that produces a chunked `BufferedRegion` per stream division. With the fix in place: 0 / 32 768 pixel diffs. With #1202 reverted locally: 28 672 / 32 768 pixels diverge, output contains NaN. Strong regression guard.

</details>

<details>
<summary>Test summary</summary>

| | |
|---|---|
| Form | New `TEST` block appended to `Modules/Filtering/ImageGrid/test/itkResampleImageFilterGTest.cxx` (no new files) |
| Geometry | 32³ float ramp, AffineTransform `Scale(0.9)`, GaussianInterp σ=1, α=1 |
| Pipeline | `Image → CastImageFilter → ResampleImageFilter → StreamingImageFilter` |
| Comparison | streamed (8 divisions) vs non-streamed (1 division), pixel-by-pixel `EXPECT_FLOAT_EQ` |
| Local timing | 12 ms |

</details>

<!--
provenance: claude-code session 2026-05-02, takeover from #1012 by @romangrothausmann (dormant since 2019)
related_files: Modules/Filtering/ImageGrid/test/itkResampleImageFilterGTest.cxx
related_issues: closes #1011 (verifies fix landed in #1202)
authorship: original draft @romangrothausmann; superseded with a different test design that actually exercises the bug
-->

romangrothausmann commented:
Since all tests succeed, it seem that just using `GaussianInterpolateImageFunction` with a test based on `itkResampleImageTest7` does not seem to trigger the problem of #1011. I also tried using 3D data instead of 2D, but then the test fails already without using `GaussianInterpolateImageFunction` probably due to the fixed tests concerning the chunk sizes. @thewtex How did you come up with these values:
https://github.com/InsightSoftwareConsortium/ITK/blob/0b588250b03da219463e6417ddf6e6d3746e8696/Modules/Filtering/ImageGrid/test/itkResampleImageTest7.cxx#L135-L137

dzenanz commented:
@romangrothausmann examining what the actual values are? The [`GetRadius` patch](https://github.com/InsightSoftwareConsortium/ITK/commit/e4003fc3dc6ec7c6e3d5782be096db567479a5d9#diff-dfa5983cf67253ce0e03aca56a207db2R135) has increased the region by 1 pixel on all sides.

romangrothausmann commented:
> examining what the actual values are?

@dzenanz What do you mean by that? 
The increase in the reagion is clear but how to get those values for a 3D case as e.g. in: https://github.com/InsightSoftwareConsortium/ITK/pull/1012/commits/3f5fd25b46d2b96cdf1122db959f9321ae8c56c6
And why does the test succeed even for `GaussianInterpolateImageFunction` (where the radius should be larger)?

dzenanz commented:
> examining what the actual values are?

Run the program under debugger and examine the values.

`GaussianInterpolateImageFunction`'s radius depends on the _sigma_ parameter. I guess the default sigma results in radius of 1, which is the same as linear interpolation.

dzenanz commented:
This PR should be revived after #1191 and #1213 are merged.

dzenanz commented:
@romangrothausmann can you rebase this PR on current master?

romangrothausmann commented:
That should be possible, I'll give it a try.

romangrothausmann commented:
3D version without `GaussianInterpolateImageFunction` (https://github.com/InsightSoftwareConsortium/ITK/pull/1012/commits/26c91e7a042e52dc0a701046652486b65b7102cb) now works for me locally. Let's see if it still does with `GaussianInterpolateImageFunction`.

romangrothausmann commented:
Oddly, on my local PC `ctest -R ResampleImageTest8` passes even though I would expect that it fails with only the changes of https://github.com/InsightSoftwareConsortium/ITK/pull/1012/commits/452bba143180952b5f27dc05690863b9315f75fa because I would expect that the 3rd dimension would demand changes to
https://github.com/InsightSoftwareConsortium/ITK/blob/452bba143180952b5f27dc05690863b9315f75fa/Modules/Filtering/ImageGrid/test/itkResampleImageTest8.cxx#L133-L138
@dzenanz How to "Run the program under debugger" with `ctest`?

dzenanz commented:
`ctest -VV` will print the exact command line for each test. That way, you can run that command line under debugger (e.g. copy command-line parameters into Visual Studio).

hjmjohnson committed:
ENH: Add streaming-equivalence test for GaussianInterpolateImageFunction

Regression guard for issue #1011 (fixed by #1202): a Gaussian-interpolated
ResampleImageFilter must produce the same output whether or not its
downstream is streamed.

Reverts of the bounding-box fix (BufferedRegion vs LargestPossibleRegion)
reproduce a 28672 / 32768-pixel divergence with NaN output values. A
CastImageFilter is inserted upstream of the resampler so the bug surfaces;
an in-memory image's BufferedRegion never shrinks under streaming, which
is why the abandoned WIP draft #1012 no longer reproduced.

Supersedes #1012 by @romangrothausmann; uses the modern GTest harness
and fits as a TEST block in itkResampleImageFilterGTest.

Co-Authored-By: Roman Grothausmann <roman.grothausmann@mh-hannover.de>

hjmjohnson commented:
Hi @romangrothausmann — I hope you don't mind, but I took the liberty of force-pushing a single commit to this branch to revive your dormant draft as a regression test for #1011.

The original three commits have been replaced (your authorship is preserved via `Co-Authored-By:` in the new commit's message). The new test is a `TEST` block appended to `itkResampleImageFilterGTest.cxx`, and inserts a `CastImageFilter` upstream of the resampler so the bug actually surfaces under streaming — with that change, locally reverting #1202's fix produces 28 672 / 32 768 NaN-or-divergent pixels, which gives us the strong regression guard you set out to add seven years ago.

cc @dzenanz — keeping this PR draft pending CI; happy to flip to ready-for-review once CI is green if you'd like.

If you'd prefer a different attribution arrangement or a separate PR from my fork, just say the word and I'll redo it.

dzenanz commented:
Let's give @romangrothausmann a few days to review this.

hjmjohnson commented:
Folded the two P2 style nits from greptile's review into the existing commit (autosquash, force-push to refresh CI). Both are pure stylistic improvements — no logic change, test still passes locally.

| Greptile P2 | Fix |
|---|---|
| Brace-aggregate `{ { 32, 32, 32 } }` for `SizeType` | Replaced with `ImageType::SizeType::Filled(32)` (matches ITK convention for uniform sizes) |
| Magic literal `8` for stream divisions | Extracted as `constexpr unsigned int kNumberOfStreamDivisions{ 8 }` |

@romangrothausmann — the patchset is still a single commit so the diff you'll see vs. `upstream/main` reflects the latest polish. No need for you to address greptile's findings yourself.

cc @dzenanz — taking the courtesy window into account; this just preempts the polish so Roman gets the final form on first read.

hjmjohnson commented on Modules/Filtering/ImageGrid/test/itkResampleImageFilterGTest.cxx:
```diff
@@ -172,3 +178,64 @@ TEST(ResampleImageFilter, ThrowsOnIncompleteConfiguration)
+TEST(ResampleImageFilter, StreamingMatchesNonStreamedWithGaussianInterpolator)
+{
+  constexpr unsigned int Dimension{ 3 };
+  using PixelType = float;
```
Folded into the squashed commit (4002ed5b7f) — `ImageType::SizeType::Filled(32)`.

hjmjohnson commented on Modules/Filtering/ImageGrid/test/itkResampleImageFilterGTest.cxx:
```diff
@@ -172,3 +178,64 @@ TEST(ResampleImageFilter, ThrowsOnIncompleteConfiguration)
+
+  streamer->SetNumberOfStreamDivisions(1);
+  ASSERT_NO_THROW(streamer->UpdateLargestPossibleRegion());
+  const ImageType::Pointer unstreamed = streamer->GetOutput();
```
Folded into the squashed commit (4002ed5b7f) — `constexpr unsigned int kNumberOfStreamDivisions{ 8 };`.
